feat: add validation layer for `metadata_properties` and `metadata_filters`

alvarobartt opened the pull request:
# Description

This PR adds a validation layer for both the `metadata_properties` when `metadata` is provided via `FeedbackRecord` when adding that record to either a local or existing `FeedbackDataset`, as well as validating the `metadata_filters` provided to the `filter_by` method for `RemoteFeedbackDataset`.

These validation layers would help reduce the API workload as the validation will be done client-side, which implies that the client will handle the validation first.

In this case, as we have the following properties: `TermsMetadataProperty`, `IntegerMetadataProperty` and `FloatMetadataProperty`; we'll need to ensure that the values provided for those are correct, both on `add_records` and also on `filter_by`.

The approach is similar to the one already implemented for the `fields` of a `FeedbackRecord`, but extending it to the `metadata_properties` to also include a validation layer that ensures not just the type of the value provided, but whether it's compliant with the pre-defined values for `TermMetadataProperty` or the pre-defined boundaries for both `IntegerMetadataProperty` and `FloatMetadataProperty`. The core difference is that besides the `pydantic` validation, there's also a method in each property that validates the values provided in the filters related to those metadata properties.

Closes #3859

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Add unit tests for the `generate_pydantic_schema_for_metadata` method
- [x] Add unit tests for both `_validate_filter` and `_pydantic_field_with_validator` in each property i.e. `TermsMetadataProperty`, `IntegerMetadataProperty`, and `FloatMetadataProperty`
- [x] Add unit tests for `add_records` with both correct and wrong `metadata`
- [x] Add integration tests for `add_records` and `filter_by` to ensure that the validation is correct both before and after applying those ops

**Checklist**

- [ ] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
refactor: rename `generate_pydantic_schema_for_fields`

alvarobartt committed:
feat: define `METADATA_PROPERTY_TYPE_TO_PYTHON_TYPE`

alvarobartt committed:
feat: add `_pydantic_field_with_validator`

alvarobartt committed:
feat: add `generate_pydantic_schema_for_metadata`

alvarobartt committed:
feat: metadata validation in `_validate_records`

frascuchon commented on src/argilla/client/feedback/schemas/metadata.py:
```diff
@@ -98,6 +103,19 @@ def check_values(cls, terms_values: List[str], values: Dict[str, Any]) -> List[s
+        if introduced_value not in self.values:
+            raise ValueError(
+                f"Provided '{self.name}={introduced_value}' is not valid, only values in {self.values} are allowed."
+            )
```
Is this a pydantic validator?

alvarobartt replied:
Well it's just a class-method in the `pydantic.BaseModel` which is later used as a validator in the property `_pydantic_field_with_validator`

alvarobartt committed:
fix(style): add missing imports under `TYPE_CHECKING`

alvarobartt committed:
feat: add `metadata_property_by_name`

alvarobartt committed:
feat: add `_validate_filter` method

alvarobartt committed:
feat: add validation layer in `filter_by`

alvarobartt committed:
test: `_validate_filter` and `_pydantic_field_with_validator`

alvarobartt committed:
docs: update `CHANGELOG.md`

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/add-metadata-validation

alvarobartt committed:
fix(style): extend type-hints for `AllowedRemote{Field,MetadataProperty}Types`

alvarobartt committed:
fix: use `pydantic` strict types

alvarobartt committed:
test: add `generate_pydantic_schema_for_metadata`

alvarobartt committed:
fix(test): add missing `__init__.py`

alvarobartt committed:
style(test): rename to `TestFeedbackDataset` and better imports

alvarobartt committed:
fix: `_pydantic_field_with_validator` default None

Because the `metadata` is always optional, so it will just be validated if provided

alvarobartt committed:
test: `add_records` for `fields` and `metadata`

alvarobartt committed:
fix(test): add missing `__init__.py`

alvarobartt committed:
fix(test): remove duplicated `FeedbackDataset.__init__`

alvarobartt committed:
test(style): remove unused imports

alvarobartt committed:
fix: `filter_by` validation on `metadata_filters`

alvarobartt committed:
test: `test_add_records` with `field` and `metadata` validation

alvarobartt committed:
test: add `test_filter_by_metadata`

alvarobartt committed:
fix(revert): raise `NotImplementedError` as `return 0` is misleading

alvarobartt committed:
fix: `filter_by` validation to allow str too

alvarobartt committed:
fix: generate `_metadata_schema` if applicable

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -177,12 +178,36 @@ def filter_by(
+        if response_status:
+            if not isinstance(response_status, list):
+                response_status = [response_status]
+            if not all(status in [arg.value for arg in FeedbackResponseStatusFilter] for status in response_status):
```
I would create a constant called `_RESPONSE_STATUS_FILTER_VALUES` or something similar with value `[arg.value for arg in FeedbackResponseStatusFilter]`

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -177,12 +178,36 @@ def filter_by(
+            if not isinstance(metadata_filters, list):
+                metadata_filters = [metadata_filters]
+            if not all(
+                metadata_filter.name in [metadata_property.name for metadata_property in self.metadata_properties]
```
I would create a variable `metadata_properties_names` just after the `if metadata_filters`

gabrielmbmb commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -213,8 +213,28 @@ def question_by_name(self, name: str) -> Union[AllowedQuestionTypes, "AllowedRem
+        """Returns the metadata properties that will be indexed and could be used to filter the dataset."""
         return self._metadata_properties
 
+    def metadata_property_by_name(
```
maybe we could cache the return value of this function

alvarobartt committed:
refactor: add `_metadata_properties_mapping`

alvarobartt committed:
fix: `_metadata_properties_mapping` if not there

alvarobartt committed:
fix: remove `warnings` from `NotImplementedError`

alvarobartt committed:
fix: wrong `AllowedRemoteMetadataPropertyTypes` type-hint

alvarobartt committed:
fix: remove deprecated `@abstractproperty`

In favour of `@property` + `@abstractmethod`. More information at https://docs.python.org/3.11/library/abc.html#abc.abstractproperty

alvarobartt committed:
test: add missing `_parse_and_validate_records`
